File: android/src/main/kotlin/com/cmc/c_shield_aip/c_shield_aip/bridges/ApBridge.kt
package com.cmc.c_shield_aip.c_shield_aip.bridges

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.cmc.c_shield_aip.c_shield_aip.CShieldErrorCode
import com.cmc.cshield_embedded.aip.api.AIPCore
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class ApBridge(private val context: Context) {

    private val mainHandler = Handler(Looper.getMainLooper())

    // Only the cryptographic sign/verify are exposed to Flutter. Body
    // normalization, payload construction, hashing and the response
    // timestamp-window check are all done in Dart (APNormalizer / CShieldAP),
    // so no fabricated okhttp Request/URL is needed here.
    fun handle(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "ap.sign"   -> handleSign(call, result)
            "ap.verify" -> handleVerify(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleSign(call: MethodCall, result: MethodChannel.Result) {
        val payload = call.argument<String>("payload")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "payload required", null)

        Thread {
            try {
                val signature = AIPCore.sign(context, payload)
                mainHandler.post { result.success(signature) }
//            } catch (e: CShieldException) {
//                mainHandler.post { result.error(CShieldErrorCode.AIP_PROXY_CA, e.message, null) }
            } catch (e: Throwable) {
                mainHandler.post { result.error(CShieldErrorCode.AP_SIGNING_FAILED, e.message, null) }
            }
        }.start()
    }

    private fun handleVerify(call: MethodCall, result: MethodChannel.Result) {
        val payload = call.argument<String>("payload")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "payload required", null)
        val signature = call.argument<String>("signature")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "signature required", null)

        Thread {
            try {
                val valid = AIPCore.verifySign(context, payload, signature)
                mainHandler.post {
                    if (valid) result.success(null)
                    else result.error(CShieldErrorCode.AP_INVALID_SIGNATURE, "Signature verification failed", null)
                }
//            } catch (e: CShieldException) {
//                mainHandler.post { result.error(CShieldErrorCode.AIP_PROXY_CA, e.message, null) }
            } catch (e: Throwable) {
                mainHandler.post { result.error(CShieldErrorCode.NATIVE_ERROR, e.message, null) }
            }
        }.start()
    }

}

File: ios/Classes/bridges/ApBridge.swift
import Flutter
import CShieldEmbedded

/// Mirrors ApBridge.kt (android) — keep both in sync.
final class ApBridge {

    // Only the cryptographic sign/verify are exposed to Flutter. Body
    // normalization, payload construction, hashing and the response
    // timestamp-window check are all done in Dart (APNormalizer / CShieldAP),
    // so no fabricated URLRequest/URL is needed here.
    func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "ap.sign":   sign(args: args, result: result)
        case "ap.verify": verify(args: args, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Sign / verify

    private func sign(args: [String: Any], result: @escaping FlutterResult) {
        guard let payload = args["payload"] as? String else { result(invalidArg()); return }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let signature = try CShieldAIP.sign(payload)
                DispatchQueue.main.async { result(signature) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: CShieldErrorCode.apSigningFailed,
                                         message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func verify(args: [String: Any], result: @escaping FlutterResult) {
        guard let payload   = args["payload"] as? String,
              let signature = args["signature"] as? String else { result(invalidArg()); return }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let valid = try CShieldAIP.verifySign(payload, signature: signature)
                DispatchQueue.main.async {
                    if valid {
                        result(nil)
                    } else {
                        result(FlutterError(code: CShieldErrorCode.apInvalidSignature,
                                             message: "Signature verification failed", details: nil))
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: CShieldErrorCode.nativeError,
                                         message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    // MARK: - Helpers

    private func invalidArg() -> FlutterError {
        FlutterError(code: CShieldErrorCode.invalidArgument, message: "Missing or invalid arguments", details: nil)
    }
}
